envoy/grpc/async_client.h

```diff
@@ -70,6 +70,20 @@ class RawAsyncStream {
    * @returns the stream info object associated with this stream.
    */
   virtual const StreamInfo::StreamInfo& streamInfo() const PURE;
+
+  /***
+   * Register a callback to be called when high/low write buffer watermark events occur on the
+   * stream. This callback must persist beyond the lifetime of the stream or be unregistered via
+   * removeWatermarkCallbacks. If there's already a watermark callback registered, this method
+   * will trigger ENVOY_BUG.
+   */
+  virtual void setWatermarkCallbacks(Http::DecoderFilterWatermarkCallbacks& callbacks) PURE;
+
+  /***
+   * Remove previously set watermark callbacks. If there's no watermark callback registered, this
+   * method will trigger ENVOY_BUG.
+   */
+  virtual void removeWatermarkCallbacks() PURE;
```

Comment:
I personally prefer to use the StreamOptions as the unique channel to propagate the context, callbacks, options between the main stream and side stream.

But considering that way may effects the basic implementation of grpc client and the http stream has used current way, so, I won't block this PR. )